consumer/retry_counter.go

```diff
@@ -0,0 +1,61 @@
+package consumer
+
+import (
+	"context"
+	"sync"
+)
+
+type retryCounter struct {
+	mu         sync.RWMutex
+	retryCount int
+
+	ch         chan State
+}
+
+func newRetryCounter() *retryCounter {
+	return &retryCounter{
+		ch : make(chan State, 2),
+	}
+}
+
+// updateLoop update the internal retry counter via notifications on its channel. It blocks until ctx is cancelled.
+func (r *retryCounter) updateLoop(ctx context.Context)  {
+	for {
+		select {
+		case state := <- r.ch:
+			r.update(state)
+		case <-ctx.Done():
+			return
```

Comment:
Could we close `ch` here instead of leaving that to the caller?

```suggestion
			close(r.ch)
			return
```

Reply:
As I understand it then normally it should be the sender that closes the channel?

Reply:
If we want to close here then we should remove the close call from the go func in the consumer?

Reply:
True, a `send` on a closed channel panics, so that makes sense. It just feels odd as it is not the owner of the channel. It is created in the constructor of `RetryCounter`. 🤔

Reply:
Yeah, so callers will just run `go rc.updateLoop(ctx)`.

Reply:
But I'm not sure of what is idiomatic TBH.

Reply:
Would it make sense to remove the channel from the ctor and create it in the consumer? Then we can pass it in to the updateLoop func

Reply:
We could do that as well. It just removes the nice encapsulation IMO. :(

Reply:
I ended up moving go routine inside again and inject the channel.

As I could not find a better way to assert the actual counter, I created a small function to check continuously for the result.